### PR TITLE
Adding the Regs3k Wayfinder

### PR DESCRIPTION
**Pull request by mistergone**
This pull request features all of the code necessary to add the "Wayfinder" feature to the Regs3k application! Here are the features:

- Implements a static wayfinder element that appears as a user browses through regs browsing pages
- Wayfinder appears when you've scrolled enough to make the first paragraph active and disappears when no paragraph is active (i.e. you're near the top of the page)
- Wayfinder shows "Regulation [LETTER]" at screen sizes when there is a left nav column visible and nothing at smaller screen sizes
- Wayfinder shows the citation of the paragraph just below it in the proper format:
  - For section pages, the citation should look like "§ 1026.5(a)(1)(ii)(A)"
  - For appendix pages, the citation should look like "Appendix E"
  - For section commentary pages, the citation should look like "Comment 5(a)(1)-2.ii"
  - For appendix commentary pages, the citation should look like "Comment app. H-1.iii.E"
  - Inline interpretations the citations should match how it looks on section commentary pages
- Wayfinder citation is a link to the active paragraph
- Wayfinder shows a note when you're in a previous or future version
  - For previous versions, the note appears as "Previous version (effective April 19, 2018 to June 1, 2018)"
  - For future versions, the note appears as "Future version (effective May 1, 2020)"
- Wayfinder appears when a page is loaded with a link hash, and the wayfinder does not obscure the paragraph it is linked to.

## Additions
- Templating, Javascript, and LESS to add the Wayfinder element to regulations browsing pages

## Testing
1. Load up the branch, and check out some regulations!
 - http://localhost:8000/policy-compliance/rulemaking/regulations/
 - Designs and details can be found in internal Platform issue GHE/#3090

## Screenshots
Screen Shot 2019-04-12 at 1 51 59 PM

Screen Shot 2019-04-12 at 1 52 23 PM

Screen Shot 2019-04-12 at 1 52 52 PM

## Notes
- Browsers have this "feature" where the viewport will scroll to the same place the user left off if they reload the page. This interferes with the "wayfinder on page load" operation. If you're testing this and you reload, you might notice the wayfinder covers the current paragraph instead of scrolling to just above it. This is not easily fixable, but I'm not sure if it is worth it considering the narrowness of the bug.
- For accessibility, the goal is mostly to keep the wayfinder from interfering with keyboard and voiceover navigation.

## Checklist
- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [x] Project documentation has been updated
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Browsers
- [ ] Chrome on desktop
- [ ] Firefox
- [ ] Safari on macOS
- [ ] Edge
- [ ] Internet Explorer 9, 10, and 11
- [ ] Safari on iOS
- [ ] Chrome on Android

### Accessibility

- [x] Keyboard friendly
- [x] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [x] Flexible from small to large screens
- [ ] No linting errors or warnings
- [x] JavaScript tests are passing

### CODE REVIEW

File: cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js [hunk 1]
```diff
@@ -74,12 +82,125 @@ const getCurrentParagraph = ( currentPosition, paragraphs ) => {
...
+  let paragraphMarker;
+
+  // For interpretations, the wayfinder should look like "Comment 4(a)-1.iv.A"
+  // Or like "Comment app. G-1.iv.A" for interpretations of appendices
+  if ( sectionTitle.indexOf( 'Comment for ' ) > -1 || label.indexOf('Interp') > -1 ) {
+    sectionFormattedTitle = 'Comment ';
+    paragraphMarker = getCommentMarker( label );
+  } else if ( sectionTitle.indexOf( 'Appendix ' ) === 0 ) {
+    // For appendices, the wayfinder should look like "Appendix A"
+    sectionFormattedTitle = sectionTitle.match( wayfinderRegex.appendixTitle )[0];
```
Comment (niqjohnson) [inline cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js]: So it turns out we have the occasional interpretation section that applies to *two* appendices at once, like `/policy-compliance/rulemaking/regulations/1026/2020-04-01/Interp-GH/`. That produces `TypeError: null is not an object (evaluating 't.match(o)[0]')` here.

File: cfgov/unprocessed/apps/regulations3k/css/reg-wayfinder.less [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+   Regulations 3000
+   Wayfinder
+   ========================================================================== */
+
+.o-regulations-wayfinder {
+    box-sizing: border-box;
+    width: 100%;
+    border-bottom: 2px solid @green;
+    position: fixed;
+    top: -80px;
```
Comment (niqjohnson) [inline cfgov/unprocessed/apps/regulations3k/css/reg-wayfinder.less]: Turns out 80px isn't always enough to fully hide the wayfinder on small screens when the wayfinder text runs to three lines (as often happens when looking at previous versions):

long-wayfinder-mobile

So we might need to up the `top` offset or calculate it dynamically to account for longer wayfinders.

**Comment by niqjohnson** [comment]
@willbarton, @Scotchester, @contolini, @cfarm: I added you as reviewers here since reviewing the code is way out of my depth (and since some of the code is what I wrote for the prototype). @mistergone mentioned that he'd kept a lot of my prototype code since it seemed to be doing the job, but he was curious to hear any suggestions for refactoring.

**Comment by willbarton** [comment]
I will recuse myself from the Python reviewing because the changes I'd be reviewing are ones I made 😁

File: cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js [hunk 3]
```diff
@@ -74,12 +82,125 @@ const getCurrentParagraph = ( currentPosition, paragraphs ) => {
...
+
+    if ( currentParagraph ) {
+      // To avoid hashed IDs, we use data-label for wayfinder text formatting
+      const currentLabel = document.getElementById( currentParagraph ).dataset.label;
+      const sectionTitle = wayfinder.dataset.section;
+      wayfinderInfo = getWayfinderInfo( currentLabel, sectionTitle );
+      paragraphMarker = wayfinderInfo.paragraphMarker;
+      sectionFormattedTitle = wayfinderInfo.formattedTitle;
+      mainContent.classList.add( 'show-wayfinder' );
+      wayfinderLink.href = '#' + currentParagraph;
```
Comment (contolini) [inline cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js]: Clicking the wayfinder's link positions the current paragraph at the top of the viewport which is behind the wayfinder. Ideally, it'd do something similar to what's on L200 and position it 60 pixels higher.

wayfinder

File: cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js [hunk 4]
```diff
@@ -1,5 +1,13 @@
 // Array that tracks paragraph positions
 let paragraphPositions;
+const wayfinderOffset = 45;
```
Comment (contolini) [inline cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js]: Per @niqjohnson's comment above, maybe calculate this dynamically?

```suggestion
const wayfinderOffset = regs3kWayfinder.scrollHeight + 10;
regs3kWayfinder.style.top = -Math.abs(wayfinderOffset) + 'px';
```

File: cfgov/regulations3k/regdown.py [hunk 5]
```diff
@@ -218,6 +218,7 @@ def run(self, parent, blocks):
             else:
                 el = util.etree.SubElement(parent, 'p')
             el.set('id', label)
+            el.set('data-label', label)
```
Comment (contolini) [inline cfgov/regulations3k/regdown.py]: FWIW it probably would have been more performant to do the label parsing in python and pass the generated paragraph labels to the DOM with this data attribute. There are parallels between the regdown logic and the JS logic in this PR. Not a big deal, though.

**Comment by mistergone** [comment]
@niqjohnson and @contolini - I've just pushed up a commit that _should_ fix the various scrolling issues with the wayfinder, including clicking on the wayfinder header link and various sizes (heights) of wayfinder in mobile views.

This problem was unfortunately very touchy, but I think it's resolved. (I'll do a bit more testing tomorrow.) Next I'll fix the "two appendices at once" bug and we should be on our way.

**Comment by mistergone** [comment]
As a note, when I went to rebase this branch, there was a big mess of conflicts. I think I resolved most of them correctly, but please let me know if anything looks awry.

File: cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js (hunk 1 above)
Comment (niqjohnson) [reply cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js]: Turns out this particular case was being caused by some unexpected, incorrect paragraph markers. But the updates to handle errors are working, and all the oddball, multi-letter interps and appendices I can find have wayfinders that are working as well as the underlying data allows now.

File: cfgov/unprocessed/apps/regulations3k/css/reg-wayfinder.less (hunk 2 above)
Comment (niqjohnson) [reply cfgov/unprocessed/apps/regulations3k/css/reg-wayfinder.less]: The wayfinder is fully hiding at all screen sizes for me now.

**Comment by mistergone** [comment]
__Thank you, @niqjohnson and @contolini !__

I added a small refactor and a new test:

[New function here.](https://github.com/cfpb/cfgov-refresh/blob/iregs-wayfinder-prototype/cfgov/unprocessed/apps/regulations3k/js/permalinks-utils.js#L25)

[New test here.](https://github.com/cfpb/cfgov-refresh/blob/iregs-wayfinder-prototype/test/unit_tests/apps/regulations3k/js/permalinks-utils-spec.js#L188)

It shouldn't affect anything, but check them out. I think we're ready to go with this PR now!

**Pull request merged**